tarorpc/taro.proto

```diff
@@ -508,8 +514,24 @@ message ProofFile {
     string genesis_point = 2;
 }
 
-message ProofVerifyResponse {
+message VerifyProofResponse {
     bool valid = 1;
+    DecodeProofResponse decoded_proof = 2;
+}
+
+message DecodeProofRequest {
+    bytes raw_proof = 1;
+    int64 proof_index = 2;
```

Comment:
Which lets you seek to a certain proof within the greater multi-proof? Does a value of -1 or 0 here mean "verify the entire thing"? Or is the idea it only does things in fragments?

Reply:
Ahh so DecodeProof will always only decode 1 proof, and same with VerifyProof

But VerifyProof will continue to verify the entire proof. Will make comments clear

Reply:
The "defaults to the latest proof" is incorrect here. Because "default" means you don't specify the value, in which case gRPC will use `0`. So by default it would verify the genesis proof (first in the file).

Reply:
@guggero doesn't this https://github.com/lightninglabs/taro/pull/181/files#diff-9bf3925007bcf9962c1a4b80fe74052ee85dda0a9b25915ddcb78b8f0ac075eaR90 create the default value of -1?

When testing, I observed it working properly to default to latest proof instead of genesis

Reply:
Yes, for the CLI. But people will also only interact with Tarod on the RPC (and will only look at the generated API docs), so the RPC description is important too.

Reply:
Ohhh yeah good point.. hmmmmm...

Will give this some thought, and happy to take suggestions.

Maybe 1 direction to go is having the index start at 1 instead of 0? Wdyt?

Reply:
No, I think that's just ugly. I think @Roasbeef suggested this already (maybe in a different context), we could use the `oneof` pattern that gRPC offers here:
```
oneof filter {
  bool all_proofs = 2;
  uint32 index = 3;
}
```

That way you can specify either `all_proofs` or set an index.

Reply:
If all_proofs was set to true, would you expect to see _all_ the proofs decoded in the response? 

Must say I'm not a fan of the oneof approach either, but I can't think of anything better yet either

Because if the user doesn't send all_proofs, it would default to false right?

Reply:
I think it can work if the bool is partialverify, wdyt? And we have the check for index > 0 as the first case in the switch.

So if a user doesn't set either, we assume we want a fullverify.

If the user only provides us an index > 0, we respect that and ignore partialverify

If a user provides us an index of 0, we will then look at partialverify to determine if they want the latest proof, or if partialverify is true, we know they want the 0 proof

Reply:
Or actually I think I need to understand oneof better https://www.tutorialspoint.com/protobuf/protobuf_compound_data_types.htm

Do I understand correctly that if _both_ fields are provided in a group by, the 2nd field will take precedence because it overwrites the first?

Reply:
Perhaps just making index optional is a route to go here? https://stackoverflow.com/questions/42622015/how-to-define-an-optional-field-in-protobuf-3

Reply:
Ok sorry for the confusion, I think I get it now.
```
oneof filter {
  bool all_proofs = 2;
  uint32 index = 3;
}
```

So we expect index to be the most common field that people would _explicitly_ send. If they do not explicitly send an index, we can default to all_proofs

```
switch filter := in.Filter.(type) {
        case *Index:
		We use the provided index and perform partial_verify

	default:
		We do a full verification
	}
```

Reply:
This is what I have

Lmk your thoughts @guggero 

```
	var proofIdx int64
	switch filter := in.Filter.(type) {
	case *tarorpc.DecodeProofRequest_ProofIndex:
		proofIdx = filter.ProofIndex
	default:
		proofIdx = -1
	}

	decodedProof, err := marshalProofFile(proofFile, proofIdx)
	if err != nil {
		return nil, fmt.Errorf("unable to marshal proof: %w", err)
	}
	```

Reply:
And then inside tarocli/proofs.go:

```
req := &tarorpc.DecodeProofRequest{
		RawProof: rawFile,
	}

	if !ctx.Bool(allProofsName) {
		req.Filter = &tarorpc.DecodeProofRequest_ProofIndex{
			ProofIndex: ctx.Int64(proofIndexName),
		}
	} else {
		req.Filter = &tarorpc.DecodeProofRequest_AllProofs{
			AllProofs: true,
		}
	}

	resp, err := client.DecodeProof(ctxc, req)
	if err != nil {
		return fmt.Errorf("unable to verify file: %w", err)
	}
	```
	
	Let me know if this works, as I've tested the solution and it works properly over REST!

Reply:
Yes, this looks good. Though I think I wanted to call it `latest_proof` instead of `all_proofs`, was answering in a rush without reloading all context, sorry.